### DISCUSSION TITLE
Define Constraints on ODE System Parameters

### DISCUSSION THREAD

**Discussion by cerkezoglu** [pnnl/neuromancer#127] [Q&A] [answered]
Hello,
Is there a way to define a constraint on the custom ode system parameter?

lets assume simple system with single input and single state

`dx = a x + b u`

where the `a` is a constant that has to be positive `a > 0` and we are dealing with a system identification problem. I create my dynamics using `ODESystem` class.

Thanks in advance.

**Comment by drgona** [comment, marked as answer]
Yes, this is possible.

You can wrap model parameters or methods in the variable class and then create constraint for the PINN loss function. 
See this example:
https://github.com/pnnl/neuromancer/blob/master/examples/ODEs/Part_7_DeepKoopman.py#L156

For simple constraints such as non-negative values or box constraints, you can use ReLU clamping
https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/slim/linear.py#L242
or sigmoid scaling method
https://github.com/pnnl/neuromancer/blob/master/src/neuromancer/modules/blocks.py#L164

For the future, this type of request is better suited for discussions rather than issues.
